### PR TITLE
fbo: Add Reserve(10) and Release(10)

### PR DESCRIPTION
**Pull request by bgly**
Signed-off-by: Bryant G. Ly <bryantly@linux.vnet.ibm.com>

### CODE REVIEW

File: file_optical.c [hunk 1]
```diff
@@ -1752,6 +1752,22 @@ static int fbo_handle_cmd(struct tcmu_device *dev, struct tcmulib_cmd *cmd)
 		return fbo_emulate_mechanism_status(dev, cdb, iovec, iov_cnt,
 						    sense);
 		break;
+	case RESERVE_10:
+		if (cdb[1] & 3)
```
Comment (mikechristie) [inline file_optical.c]: I might be looking at the wrong SPC version, but is bit 3 just reserved?

File: file_optical.c [hunk 2]
```diff
@@ -1752,6 +1752,22 @@ static int fbo_handle_cmd(struct tcmu_device *dev, struct tcmulib_cmd *cmd)
...
+		break;
+	case RELEASE_10:
+		if (cdb[1] & 3)
+			return tcmu_set_sense_data(sense, ILLEGAL_REQUEST,
+						   ASC_INVALID_FIELD_IN_CDB,
+						   NULL);
+		else
+			return SAM_STAT_GOOD;
+		break;
 	default:
```
Comment (mikechristie) [inline file_optical.c]: Did part of your commit get dropped? IIs there supposed to be more code for this commit? All I see is the above code. I did not see anything like some code checking for conflicts and existing reservations.

**Comment by bgly** [comment]
> *@mikechristie* commented on this pull request.
>
> ------------------------------------------------------------------------
>
> In file_optical.c 
> <https://github.com/open-iscsi/tcmu-runner/pull/105#discussion_r106706064>:
>
> > @@ -1752,6 +1752,22 @@ static int fbo_handle_cmd(struct tcmu_device *dev, struct tcmulib_cmd *cmd)
>   		return fbo_emulate_mechanism_status(dev, cdb, iovec, iov_cnt,
>   						    sense);
>   		break;
> +	case RESERVE_10:
> +		if (cdb[1] & 3)
>
> I might be looking at the wrong SPC version, but is bit 3 just reserved?
>
>
Spec states that if the LONGID bit is one and parameter list length is less than eight, the device should return check condition with illegal request.
It also states that if Byte 1, bit 0 is equal to one, thedevice servers not implementing the obsolete capability shall terminate the command
with CHECK CONDITION status and the sense key shall be set to ILLEGAL REQUEST.



-Bryant

**Comment by bgly** [comment]
I'm not sure how to fail commands for fail commands from initiators other than the one who sent the Reserve, but at this level I don't know anything about the initiator? -- If you can help Mike

**Comment by mikechristie** [comment]
I am not sure yet. I have been thinking about this for full PGR support.

For ALUA support we had a similar problem where we need to check if the command is executable on the path it was received on.

Here was my RFC patch for ALUA:
https://www.spinics.net/lists/target-devel/msg13999.html

We have tcmu-runner handle commands like STPG and RTPG by writing/reading to the LIO configfs interface to set/get ALUA state. We then rely on LIO core in the kernel to do the state/path checks for each command. Here are the runner patches that would go with it:

https://github.com/mikechristie/tcmu-runner/commits/alua-kern-setup

For PGRs, we could do something similar where we have runner process the PGR related commands and then it interacts with the kernel to update LIO core's state and we rely on the kernel's __target_execute_cmd()->target_check_reservation() check.

The kernel has code to read/write the PGR state from configfs but I have not looked into how easy it will be easy to use.

**Comment by bgly** [comment]
Hi Mike and Nick,

I am trying to handle Reservation commands within tcmu-runner but its hard since it doesnt know anything about the initiator.

So looking at kernel code... Can we just have LIO handle all the reserves? Reserve/Persistent Reserves?

within:

sense_reason_t

passthrough_parse_cdb(struct se_cmd *cmd,

         sense_reason_t (*exec_cmd)(struct se_cmd *cmd))

Have this function handle PERSISTENT_RESERVE_IN:, PERSISTENT_RESERVE_OUT:, RELEASE:/RESERVE:(10).
It looks like LIO already handles REPORT_LUNS there no matter what, so I think these can also be added to that list?

I cant see how a pass-through device would know anything about the initiator so its not like they can handle these reserve commands anyways?


What do you think Nick/Mike?


-Bryant Ly

**Comment by bgly** [comment]
On 03/17/2017 02:07 PM, Bryant G. Ly wrote:
> Hi Mike and Nick, 
> 
> I am trying to handle Reservation commands within tcmu-runner but its hard since it doesnt know anything about the initiator. 
> 
> So looking at kernel code... Can we just have LIO handle all the reserves? Reserve/Persistent Reserves?
> 
> within: 
> 
> sense_reason_t
> 
> passthrough_parse_cdb(struct se_cmd *cmd,
> 
>         sense_reason_t (*exec_cmd)(struct se_cmd *cmd))
> 
> Have this function handle PERSISTENT_RESERVE_IN:, PERSISTENT_RESERVE_OUT:, RELEASE:/RESERVE:(10).
> It looks like LIO already handles REPORT_LUNS there no matter what, so I think these can also be added to that list?
> 
> I cant see how a pass-through device would know anything about the initiator so its not like they can handle these reserve commands anyways?
> 
> 
> What do you think Nick/Mike?
> 

For HA we need the PGR command or the PGR info sent to userspace, so we
can distribute the info across nodes.

We can either handle all that by implementing it like we did for ALUA
where tcmu-runner processes the commands, and then writes/reads to configfs.

Or,

The kernel can process the commands and we have to tell userspace about
state changes like new registration/reservation info through something
like netlink.

It does not matter to me. I think it would be best to handle ALUA and
PGRs that same though.

**Comment by bgly** [comment]
I agree, we should do it like you did for ALUA. I'll have to review that RFC patch.

**Comment by bgly** [comment]
On 03/17/2017 02:15 PM, Mike Christie wrote:
> On 03/17/2017 02:07 PM, Bryant G. Ly wrote:
>> Hi Mike and Nick, 
>>
>> I am trying to handle Reservation commands within tcmu-runner but its hard since it doesnt know anything about the initiator. 
>>
>> So looking at kernel code... Can we just have LIO handle all the reserves? Reserve/Persistent Reserves?
>>
>> within: 
>>
>> sense_reason_t
>>
>> passthrough_parse_cdb(struct se_cmd *cmd,
>>
>>         sense_reason_t (*exec_cmd)(struct se_cmd *cmd))
>>
>> Have this function handle PERSISTENT_RESERVE_IN:, PERSISTENT_RESERVE_OUT:, RELEASE:/RESERVE:(10).
>> It looks like LIO already handles REPORT_LUNS there no matter what, so I think these can also be added to that list?
>>
>> I cant see how a pass-through device would know anything about the initiator so its not like they can handle these reserve commands anyways?
>>
>>
>> What do you think Nick/Mike?
>>
> 
> For HA we need the PGR command or the PGR info sent to userspace, so we
> can distribute the info across nodes.
> 
> We can either handle all that by implementing it like we did for ALUA
> where tcmu-runner processes the commands, and then writes/reads to configfs.
> 

I think for PGR we have to do what Bryant is suggesting above where the
kernel processes the commands.

The problem is that for most apps are relying on the I_T nexues that the
register is sent on getting registered. They are not sending ports or
setting the all target ports bit.

We could split ALUA because it was sending the target port group and
relative port ids in the STPG and RTPG are supposed to report all port
groups the device is using, so we had enough info in userspace.

If we want to process PGRs in userspace we have to do something like tgt
where the kernel part passed tgtd a event with a I_T nexues info and
special ID and each command that passed to userspace got a I_T nexus ID.


> Or,
> 
> The kernel can process the commands and we have to tell userspace about
> state changes like new registration/reservation info through something
> like netlink.
> 
> It does not matter to me. I think it would be best to handle ALUA and
> PGRs that same though.
>

**Comment by bgly** [comment]
On 3/17/17 6:14 PM, Mike Christie wrote:
> On 03/17/2017 02:15 PM, Mike Christie wrote:
>> On 03/17/2017 02:07 PM, Bryant G. Ly wrote:
>>> Hi Mike and Nick,
>>>
>>> I am trying to handle Reservation commands within tcmu-runner but its hard since it doesnt know anything about the initiator.
>>>
>>> So looking at kernel code... Can we just have LIO handle all the reserves? Reserve/Persistent Reserves?
>>>
>>> within:
>>>
>>> sense_reason_t
>>>
>>> passthrough_parse_cdb(struct se_cmd *cmd,
>>>
>>>          sense_reason_t (*exec_cmd)(struct se_cmd *cmd))
>>>
>>> Have this function handle PERSISTENT_RESERVE_IN:, PERSISTENT_RESERVE_OUT:, RELEASE:/RESERVE:(10).
>>> It looks like LIO already handles REPORT_LUNS there no matter what, so I think these can also be added to that list?
>>>
>>> I cant see how a pass-through device would know anything about the initiator so its not like they can handle these reserve commands anyways?
>>>
>>>
>>> What do you think Nick/Mike?
>>>
>> For HA we need the PGR command or the PGR info sent to userspace, so we
>> can distribute the info across nodes.
>>
>> We can either handle all that by implementing it like we did for ALUA
>> where tcmu-runner processes the commands, and then writes/reads to configfs.
>>
> I think for PGR we have to do what Bryant is suggesting above where the
> kernel processes the commands.
>
> The problem is that for most apps are relying on the I_T nexues that the
> register is sent on getting registered. They are not sending ports or
> setting the all target ports bit.
>
> We could split ALUA because it was sending the target port group and
> relative port ids in the STPG and RTPG are supposed to report all port
> groups the device is using, so we had enough info in userspace.
>
> If we want to process PGRs in userspace we have to do something like tgt
> where the kernel part passed tgtd a event with a I_T nexues info and
> special ID and each command that passed to userspace got a I_T nexus ID.
>
>
So a patch along the lines of :

Bryants-MacBook-Pro:ibmvscsis bgly$ git diff drivers/target/target_core_device.c
diff --git a/drivers/target/target_core_device.c b/drivers/target/target_core_device.c
index f75da7a..1b82289 100644
--- a/drivers/target/target_core_device.c
+++ b/drivers/target/target_core_device.c
@@ -1088,6 +1088,29 @@ passthrough_parse_cdb(struct se_cmd *cmd,
                 return TCM_NO_SENSE;
         }

+       /*
+        * For PERSISTENT RESERVE IN/OUT, RELEASE, and RESERVE we need to
+        * emulate the response, since pass throughs dont have the information
+        * required to process these commands.
+        */
+       if (cdb[0] == PERSISTENT_RESERVE_IN) {
+               cmd->execute_cmd = target_scsi3_emulate_pr_in;
+               return TCM_NO_SENSE;
+       }
+       if (cdb[0] == PERSISTENT_RESERVE_OUT) {
+               cmd->execute_cmd = target_scsi3_emulate_pr_out;
+               return TCM_NO_SENSE;
+       }
+
+       if (cdb[0] == RELEASE || cdb[0] == RELEASE_10) {
+               cmd->execute_cmd = target_scsi2_reservation_release;
+               return TCM_NO_SENSE;
+       }
+       if (cdb[0] == RESERVE || cdb[0] == RESERVE_10) {
+               cmd->execute_cmd = target_scsi2_reservation_reserve;
+               return TCM_NO_SENSE;
+       }
+
         /* Set DATA_CDB flag for ops that should have it */
         switch (cdb[0]) {
         case READ_6:

Bryants-MacBook-Pro:ibmvscsis bgly$ git diff drivers/target/target_core_pr.c
diff --git a/drivers/target/target_core_pr.c b/drivers/target/target_core_pr.c
index 47463c9..e38aedc 100644
--- a/drivers/target/target_core_pr.c
+++ b/drivers/target/target_core_pr.c
@@ -4146,8 +4146,6 @@ target_check_reservation(struct se_cmd *cmd)
                 return 0;
         if (dev->se_hba->hba_flags & HBA_FLAGS_INTERNAL_USE)
                 return 0;
-       if (dev->transport->transport_flags & TRANSPORT_FLAG_PASSTHROUGH)
-               return 0;

         spin_lock(&dev->dev_reservation_lock);
         if (dev->dev_reservation_flags & DRF_SPC2_RESERVATIONS)

Do you think something like this would be suffice?

-Bryant

**Comment by bgly** [comment]
On 03/20/2017 11:21 AM, Bryant G. Ly wrote:
> 
> 
> On 3/17/17 6:14 PM, Mike Christie wrote:
>> On 03/17/2017 02:15 PM, Mike Christie wrote:
>>> On 03/17/2017 02:07 PM, Bryant G. Ly wrote:
>>>> Hi Mike and Nick,
>>>>
>>>> I am trying to handle Reservation commands within tcmu-runner but
>>>> its hard since it doesnt know anything about the initiator.
>>>>
>>>> So looking at kernel code... Can we just have LIO handle all the
>>>> reserves? Reserve/Persistent Reserves?
>>>>
>>>> within:
>>>>
>>>> sense_reason_t
>>>>
>>>> passthrough_parse_cdb(struct se_cmd *cmd,
>>>>
>>>>          sense_reason_t (*exec_cmd)(struct se_cmd *cmd))
>>>>
>>>> Have this function handle PERSISTENT_RESERVE_IN:,
>>>> PERSISTENT_RESERVE_OUT:, RELEASE:/RESERVE:(10).
>>>> It looks like LIO already handles REPORT_LUNS there no matter what,
>>>> so I think these can also be added to that list?
>>>>
>>>> I cant see how a pass-through device would know anything about the
>>>> initiator so its not like they can handle these reserve commands
>>>> anyways?
>>>>
>>>>
>>>> What do you think Nick/Mike?
>>>>
>>> For HA we need the PGR command or the PGR info sent to userspace, so we
>>> can distribute the info across nodes.
>>>
>>> We can either handle all that by implementing it like we did for ALUA
>>> where tcmu-runner processes the commands, and then writes/reads to
>>> configfs.
>>>
>> I think for PGR we have to do what Bryant is suggesting above where the
>> kernel processes the commands.
>>
>> The problem is that for most apps are relying on the I_T nexues that the
>> register is sent on getting registered. They are not sending ports or
>> setting the all target ports bit.
>>
>> We could split ALUA because it was sending the target port group and
>> relative port ids in the STPG and RTPG are supposed to report all port
>> groups the device is using, so we had enough info in userspace.
>>
>> If we want to process PGRs in userspace we have to do something like tgt
>> where the kernel part passed tgtd a event with a I_T nexues info and
>> special ID and each command that passed to userspace got a I_T nexus ID.
>>
>>
> So a patch along the lines of :
> 
> Bryants-MacBook-Pro:ibmvscsis bgly$ git diff
> drivers/target/target_core_device.c
> diff --git a/drivers/target/target_core_device.c
> b/drivers/target/target_core_device.c
> index f75da7a..1b82289 100644
> --- a/drivers/target/target_core_device.c
> +++ b/drivers/target/target_core_device.c
> @@ -1088,6 +1088,29 @@ passthrough_parse_cdb(struct se_cmd *cmd,
>                 return TCM_NO_SENSE;
>         }
> 
> +       /*
> +        * For PERSISTENT RESERVE IN/OUT, RELEASE, and RESERVE we need to
> +        * emulate the response, since pass throughs dont have the
> information
> +        * required to process these commands.
> +        */
> +       if (cdb[0] == PERSISTENT_RESERVE_IN) {
> +               cmd->execute_cmd = target_scsi3_emulate_pr_in;
> +               return TCM_NO_SENSE;
> +       }
> +       if (cdb[0] == PERSISTENT_RESERVE_OUT) {
> +               cmd->execute_cmd = target_scsi3_emulate_pr_out;
> +               return TCM_NO_SENSE;
> +       }
> +
> +       if (cdb[0] == RELEASE || cdb[0] == RELEASE_10) {
> +               cmd->execute_cmd = target_scsi2_reservation_release;
> +               return TCM_NO_SENSE;
> +       }
> +       if (cdb[0] == RESERVE || cdb[0] == RESERVE_10) {
> +               cmd->execute_cmd = target_scsi2_reservation_reserve;
> +               return TCM_NO_SENSE;
> +       }
> +
>         /* Set DATA_CDB flag for ops that should have it */
>         switch (cdb[0]) {
>         case READ_6:
> 
> Bryants-MacBook-Pro:ibmvscsis bgly$ git diff
> drivers/target/target_core_pr.c
> diff --git a/drivers/target/target_core_pr.c
> b/drivers/target/target_core_pr.c
> index 47463c9..e38aedc 100644
> --- a/drivers/target/target_core_pr.c
> +++ b/drivers/target/target_core_pr.c
> @@ -4146,8 +4146,6 @@ target_check_reservation(struct se_cmd *cmd)
>                 return 0;
>         if (dev->se_hba->hba_flags & HBA_FLAGS_INTERNAL_USE)
>                 return 0;
> -       if (dev->transport->transport_flags & TRANSPORT_FLAG_PASSTHROUGH)
> -               return 0;
> 
>         spin_lock(&dev->dev_reservation_lock);
>         if (dev->dev_reservation_flags & DRF_SPC2_RESERVATIONS)
> 
> Do you think something like this would be suffice?
> 

Yeah, you would also need to convert tagret_core_configfs.c so you can
get/set the PR info and if we wanted to restrict pscsi we would need to
add a new transport_flag like was done for ALUA.

Nick are you ok with this approach though? It sounds like Christoph does
not like it. If you are also against it, then there is no point in
Bryant making the changes I mention.

If we do not go this route, then the only alternative I can think of is
to do tgt style where the kernel/lio passes userspace I_T Nexus info.

**Pull request closed without merging**